discojs/src/models/gpt/layers.spec.ts

```diff
@@ -0,0 +1,308 @@
+
+      // expected output shape for 2D input: [2, 5, nEmbd]
+      expect(output.shape).to.deep.equal([2, 5, nEmbd]);
+    });
+
+    it('should return token logits with shape [batch_size, sequence_length, vocabSize] for 3D input', function () {
+      const vocabSize = 100;
+      const nEmbd = 16;
+      const seed = 42;
+
+      const lmEmbedding = new LMEmbedding(vocabSize, nEmbd, seed);
+
+      // dummy 3D input representing a batch of embeddings: shape [batch_size, sequence_length, nEmbd]
+      const embeddingsInput = tf.randomUniform([2, 5, nEmbd]);
+
+      const output = lmEmbedding.apply(embeddingsInput) as tf.Tensor;
+
+      // expected output shape for 3D input: [2, 5, vocabSize]
+      expect(output.shape).to.deep.equal([2, 5, vocabSize]);
+    });
+
+    it('should throw an error for unexpected input shape', function () {
+      const vocabSize = 100;
+      const nEmbd = 16;
+      const seed = 42;
+
+      const lmEmbedding = new LMEmbedding(vocabSize, nEmbd, seed);
+
+      // invalid input tensor with 1D shape.
+      const invalidInput = tf.tensor1d([1, 2, 3], 'int32');
+
+      expect(() => lmEmbedding.apply(invalidInput)).to.throw('unexpected input shape');
+    });
+
+    it('should throw an error if input is an array with more than one tensor', function () {
+      const vocabSize = 100;
+      const nEmbd = 16;
+      const seed = 42;
+      const lmEmbedding = new LMEmbedding(vocabSize, nEmbd, seed);
+      const input1 = tf.tensor2d([[1, 2, 3]], [1, 3], 'int32');
+      const input2 = tf.tensor2d([[4, 5, 6]], [1, 3], 'int32');
+      expect(() => lmEmbedding.apply([input1, input2])).to.throw('expected exactly one tensor');
+    });
+
+    it('should compute correct output shape for 2D input using computeOutputShape', function () {
+      const vocabSize = 100;
+      const nEmbd = 16;
+      const seed = 42;
+      const lmEmbedding = new LMEmbedding(vocabSize, nEmbd, seed);
+      const outputShape = lmEmbedding.computeOutputShape([null, null]);
+      expect(outputShape).to.deep.equal([null, null, nEmbd]);
+    });
+
+  });
+
+  // Range Layer tests
+  describe('Range Layer', function () {  
+
+    afterEach(() => {
+      // dispose any created tensors/variables
+      tf.disposeVariables();
+    });
+  
+    it('should output a tensor with shape [1, T] for an input of shape [batch, T]', async function () {
+      const rangeLayer = new Range();
+  
+      // dummy input tensor with shape [batch, T]
+      const dummyInput = tf.zeros([3, 10], 'int32');
+  
+      const output = rangeLayer.apply(dummyInput) as tf.Tensor;
+  
+      // We expect the output to have shape [1, T] i.e. [1, 10]
+      expect(output.shape).to.deep.equal([1, 10]);
+  
+      // verify the content: the layer should output a range [0, 1, ..., T-1]
+      const outputData = await output.data();
+      for (let i = 0; i < 10; i++) {
+        expect(outputData[i]).to.equal(i);
+      }
```

Comment:
for small arrays, I find that checking directly is more readable

```suggestion
      expect(await output.data()).to.deep.equal(
        Int32Array.of(0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
      );
```